pkg/services/ngalert/state/manager.go

```diff
@@ -314,41 +312,79 @@ func (st *Manager) flushState(ctx context.Context) {
 	st.cache.mtxStates.Lock()
 	defer st.cache.mtxStates.Unlock()
 	totalStates, errorsCnt := 0, 0
+	var stateSlice []*State
 	for _, orgStates := range st.cache.states {
 		for _, ruleStates := range orgStates {
+			stateSlice = stateSlice[:0]
 			for _, state := range ruleStates {
-				err := st.saveState(ctx, state)
-				totalStates++
-				if err != nil {
-					st.log.Error("failed to save alert state", append(state.GetRuleKey().LogContext(), "labels", state.Labels.String(), "state", state.State.String(), "err", err.Error()))
-					errorsCnt++
-				}
+				stateSlice = append(stateSlice, state)
+			}
+			savedCount, failedCount := st.saveAlertStates(ctx, stateSlice...)
+			if failedCount != 0 {
+				st.log.Error("failed to save alert states.", "count", failedCount)
 			}
+			errorsCnt += failedCount
+			totalStates += savedCount
```

Comment:
Any reason why you decided to do it by rule?I wonder if we can just dump all states at once. Separating by rule won't help here because there can be a rule that has 1000s of states and a rule that has just 1 state.

Reply:
I was trying to mimic the logging behavior of the code that was already in place. That's probably not all that important to preserve.

I think we could set a reasonably large batch size instead of doing this rule-by-rule, just so we don't allocate an unbounded set of data. (I understand that the code that's already here is unbounded, I should fix that)